### PR TITLE
add prometheus Gauge Channel Length

### PR DESCRIPTION
**Pull request by amitbasuri**
Resolves https://github.com/katzenpost/katzenpost/issues/574

### CODE REVIEW

File: server/internal/gateway/gateway.go [hunk 1]
```diff
@@ -311,6 +311,18 @@ func New(glue glue.Glue) (glue.Gateway, error) {
...
+
 	isOk = true
 	return p, nil
 }
+
+func (p *gateway) monitorChannelLen() {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+
+	for range ticker.C {
```
Comment (david415) [inline server/internal/gateway/gateway.go:325]: We've got a worker type and this `gateway` type is composite with the `Worker` type:
https://github.com/katzenpost/katzenpost/blob/main/server/internal/gateway/gateway.go#L38
https://github.com/katzenpost/katzenpost/blob/main/core/worker/worker.go#L23

And so... a very common pattern we use inside the mix server is to maintain a server shutdown code path:

```golang
select {
        case <-p.HaltCh():
                return
        case <-timer.C:
                // blah blah
}
```

I'm not saying we need to do this here necessarily unless the mix server no longer shutdown down cleanly.

File: server/internal/instrument/prometheus.go [hunk 2]
```diff
@@ -167,6 +169,31 @@ var (
 	)
 )
 
+var monitoredChannels = struct {
+	sync.Mutex
+	channels map[string]chan interface{}
```
Comment (david415) [inline server/internal/instrument/prometheus.go:174]: there's no need to create a struct type and store channels in a map.

File: server/internal/instrument/prometheus.go [hunk 3]
```diff
@@ -167,6 +169,31 @@ var (
...
+}{
+	channels: make(map[string]chan interface{}),
+}
+
+func startChannelLenMonitor() {
```
Comment (david415) [inline server/internal/instrument/prometheus.go:179]: this isn't ideal. we do not need nested for loops or mutex contention or a map storing things.

Comment (david415) [reply server/internal/instrument/prometheus.go]: i fixed this by pushing another commit on this branch

**Comment by david415** [comment]
I pushed a commit to this branch that simplifies the prometheus monitoring of buffered channels:

```
func MonitorChannelLen(name string, haltCh <-chan interface{}, ch chan interface{}) {
	go doMonitorChannelLen(name, haltCh, ch)
}

func doMonitorChannelLen(name string, haltCh <-chan interface{}, ch chan interface{}) {
	ticker := time.NewTicker(time.Second)
	defer ticker.Stop()
	for {
		select {
		case <-haltCh:
			return
		case <-ticker.C:
			channelUsage.With(prometheus.Labels{"channel_name": name}).Set(float64(len(ch)))
		}
	}
}
```

**Comment by amitbasuri** [comment]
@david415 yes now it looks better

**Review by david415 — APPROVED** [review]

**Pull request merged**